### ISSUE TITLE
Constraining the Solution in Prioritizr

### ISSUE THREAD

**Issue by robert-ed-elliott** [prioritizr/prioritizr#326]
Hello/Good Afternoon,  I would like to know if it's possible to constrain the solution somehow so that a single conservation unit is produced.  What I would like to see is the best solution where a minimized area may not be that useful, but rather a solution with perhaps a somewhat larger or much larger area with planning units of no value included within a single conservation unit solution. Hope I'm clear what I want to do.  The objective really is to produce a single conservation zone that includes planning units with certain values surrounded by planning units with that have no value, basically maximizing the sum of the planning units with the solution with the total value of all planning units above or at a set value.  Thanks, Ted

**Comment by jeffreyhanson** [comment]
Hi @robert-ed-elliott,

Thanks for reaching out! 

Just to clarify, when you mean "a single conservation unit", do you mean (1) a single planning unit, or (2) a set of planning units that are arranged to form a single contiguous spatial area (e.g. a giant circle)? If the latter, then I would suggest adding boundary penalties (https://prioritizr.net/reference/add_boundary_penalties.html) to the problem and selecting a suitable `penalty` value that spatially clusters all the planning units togeather (see https://prioritizr.net/articles/calibrating_trade-offs_tutorial.html for suggestions on how to do this). Another option could be to use the contiguity constraints (https://prioritizr.net/reference/add_contiguity_constraints.html), but these are very mathemematically complex and, as such, require the Gurobi solver and only solve within a reasonable time period for fairly small problems (e.g. <10k planning units). 

How does that sound? Please let me know if you have any follow up questions?

**Comment by robert-ed-elliott** [comment]
Hi Jeff.  Thanks for the reply.  I was just reading a lot of the traffic between you and Teagan regarding "Add max area threshold for zone and dealing with fragmented zones".  It sounds like they may have been dealing with a similar issue as me.  I'll create an visual example of what I was trying get across in as few words as possible.  Do you mind if I get back to you shortly.  Thanks, Ted

**Comment by jeffreyhanson** [comment]
No worries! Ah, yeah, I think Teagan ended up resolving this using connectivity penalties (based on planning unit adjcency data to promote the allocation of neighboring planning units to the same zone) and playing with inter-zone penalties (using the `zone` parameter, to discourage certain zones from being allocated to neighobring planning units -- but it sounds like you have a single zone problem, so this isn't needed). Yea, if you can provide a visual example that would be really helpful - thanks!

**Comment by robert-ed-elliott** [comment]
Hi Jeff,

What I would like to do is be able to generate a solution like the blue polygon in this figure, a single zone.  The white cells have values of 0, the light gray cells have values of 33, the medium gray cells have values of 66, and the black cells have values of 100.  The constraint/target to create the zone would be a single zone, where the cells within it sum to less than or equal to a certain value.  The tests I've done so far will not create a single zone but rather a very fragmented solution, which I don't want because the zone must not be fragmented at all for ease of operations.  In addition, if the fragmented solution is simply used to create a convex hull it will include a bunch of cells that were not in the solution it found.

Thanks,  Ted

![image](https://github.com/prioritizr/prioritizr/assets/58744885/d0c019dc-f937-45c6-b86c-6a52c83bd527)

**Comment by jeffreyhanson** [comment]
Hi, 

Thanks for providing these details! If I understand correctly, it sounds like you have a single feature and the figure you attached is showing the spatial distribution of the values for the feature (e.g. with values of 0, 33, 66, and 100). To help reduce spatial fragmentation in the solution, you could try specifying some spatial penalties or constraints. For example, you could use the boundary penalties (https://prioritizr.net/reference/add_boundary_penalties.html). That said, it looks like the spatial distribution of the single feature is very spatially fragmented (e.g. the black and grey pixels are pretty far apart from each other, and mostly surrounded by white pixels). So, you might need to set a relatively high boundary penalty value to get a solution that meets your criteria (see this tutorial for advice on calibrating the penalty value, https://prioritizr.net/articles/calibrating_trade-offs_tutorial.html). How does that sound?

**Comment by robert-ed-elliott** [comment]
Hi Jeff, Thanks for responding.  Thought I'd ask that question because I was not sure if there was another "workaround" aside from the boundary penalties you mention, things like hard rules for the number of zones that will be created for a solution.  You are right about the sparse nature of the cells with values.  Another solution I was considering was making the cells larger so there is not as much white space. My feeling is that prioritzr and other tools work well with continuous surfaces rather than binary (yes, no) cells which is what I've got here.   Cheers, Ted

**Comment by jeffreyhanson** [comment]
Yeah, generally, these tools (e.g. see also Marxan and Zonation) are designed for problems where you have many/multiple features (e.g. species, ecosystem types, ecosystem services, or surfaces describing the spatial distribution of different elements of biodiversity). They do work well with both binary and continuous data/surfaces, but it would probably be more useful to have many more features/surfaces if you only have binary data compared with continuous  data. Yeah, making the cells (planning units) larger could be one potential fix - but it might not be very satisfying (e.g., might require playing around with a range of different resolutions to find the right blance, and even then might not be ideal). Also, using a continuous cost data/layer/surface can also somewhat help with spatial clustering too (but this is more of an emergent property based on the level of spatial auto-correlation in the data, rather than something you can directly control). Note that the cost data don't strictly need to be economic ($$$) measures of cost, and you can use surrogates instead (e.g., based on measures of land degredation, human pressure, inverse nature intactness, human population density, suitability of land for agriculture, etc). Does that help?

**Comment by jeffreyhanson** [comment]
I think this issue has been resolved, so I'll close this now? Please feel free to re-open or continue commenting if I've got this wrong? Happy to keep chatting if there's anything I can help with.
